bin/faucet/src/client.rs

```diff
@@ -0,0 +1,382 @@
+        response.into_inner().block_header.unwrap().try_into().map_err(|err| {
+            FaucetError::InternalServerError(format!("Failed to convert block header: {}", err))
+        })?;
+
+    let root_chain_mmr = ChainMmr::new(
+        PartialMmr::from_peaks(
+            MmrPeaks::new(0, Vec::new()).expect("Empty MmrPeak should be valid"),
+        ),
+        Vec::new(),
+    )
+    .expect("Empty ChainMmr should be valid");
+
+    Ok((rpc_api, root_block_header, root_chain_mmr))
+}
+
+fn build_transaction_arguments(
+    output_note: &Note,
+    executor: &TransactionExecutor<FaucetDataStore, FaucetAuthenticator>,
+    note_type: NoteType,
+    asset: FungibleAsset,
+) -> Result<TransactionArgs, FaucetError> {
+    let recipient = output_note
+        .recipient()
+        .digest()
+        .iter()
+        .map(|x| x.as_int().to_string())
+        .collect::<Vec<_>>()
+        .join(".");
+
+    let tag = output_note.metadata().tag().inner();
+
+    let script = ProgramAst::parse(
+        &DISTRIBUTE_FUNGIBLE_ASSET_SCRIPT
+            .replace("{recipient}", &recipient)
+            .replace("{note_type}", &Felt::new(note_type as u64).to_string())
+            .replace("{tag}", &Felt::new(tag.into()).to_string())
+            .replace("{amount}", &Felt::new(asset.amount()).to_string()),
+    )
+    .expect("shipped MASM is well-formed");
+
+    let script = executor.compile_tx_script(script, vec![], vec![]).map_err(|err| {
+        FaucetError::InternalServerError(format!("Failed to compile script: {}", err))
+    })?;
+
+    let mut transaction_args = TransactionArgs::new(Some(script), None, AdviceMap::new());
+    transaction_args.extend_expected_output_notes(vec![output_note.clone()]);
+
+    Ok(transaction_args)
+}
+
+// TODO: Remove the falcon signature function once it's available on base and made public
+
+/// Retrieves a falcon signature over a message.
+/// Gets as input a [Word] containing a secret key, and a [Word] representing a message and
+/// outputs a vector of values to be pushed onto the advice stack.
+/// The values are the ones required for a Falcon signature verification inside the VM and they are:
+///
+/// 1. The nonce represented as 8 field elements.
+/// 2. The expanded public key represented as the coefficients of a polynomial of degree < 512.
+/// 3. The signature represented as the coefficients of a polynomial of degree < 512.
+/// 4. The product of the above two polynomials in the ring of polynomials with coefficients
+/// in the Miden field.
+///
+/// # Errors
+/// Will return an error if either:
+/// - The secret key is malformed due to either incorrect length or failed decoding.
+/// - The signature generation failed.
+///
+/// TODO: once this gets made public in miden base, remve this implementation and use the one from
+/// base
+fn get_falcon_signature(
+    key: &rpo_falcon512::SecretKey,
+    message: Word,
+    rng: &mut ChaCha20Rng,
+) -> Result<Vec<Felt>, AuthenticationError> {
```

Comment:
I believe it was mentioned in one of the comments in `minde-base` that we can get rid of this function and use `BasicTransactionAuthenticator` instead.

Reply:
Done.